### ISSUE TITLE
770 line incomprehensible error message

### ISSUE THREAD

**Issue by davidak** [NixOS/nix#7898]
**Describe the bug**

I want to apply for the award for the worst error message with this one.

https://gist.github.com/davidak/e91d0eb63c002ca0afcc5d2d75fa8432

The error message is about 770 lines long, mentions files i haven't touched and error at the end does not make sense to me.

>error: evaluation aborted with the following error message: 'unsupported platform for the pure Linux stdenv'

I used `--system i686-linux`, so that is not supported?

@trofi said in chat:

>It tells that nixpkgs does not have i686-linux bootstrap binaries for musl libc.

**Steps To Reproduce**

This produces the same error, easier, but the messages are ordered the opposite way for some reason. The error is on the top instead of the bottom.

`nix eval -f '<nixpkgs>' pkgsMusl.hello --argstr system i686-linux --show-trace`

**Expected behavior**

Short and clear error message i can act on.

**`nix-env --version` output** nix-env (Nix) 2.11.1

**Additional context**

It's fun rebuilding every package from staging.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by trofi** [comment]
This error message comes from `nixpkgs`, not `nix`. You can extend it to contain more details: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/linux/default.nix#L63-L94

```nix
  let table = {
    glibc = {
      i686-linux = import ./bootstrap-files/i686.nix;
      x86_64-linux = import ./bootstrap-files/x86_64.nix;
      armv5tel-linux = import ./bootstrap-files/armv5tel.nix;
      armv6l-linux = import ./bootstrap-files/armv6l.nix;
      armv7l-linux = import ./bootstrap-files/armv7l.nix;
      aarch64-linux = import ./bootstrap-files/aarch64.nix;
      mipsel-linux = import ./bootstrap-files/loongson2f.nix;
      mips64el-linux = import ./bootstrap-files/mips64el.nix;
      powerpc64le-linux = import ./bootstrap-files/powerpc64le.nix;
      riscv64-linux = import ./bootstrap-files/riscv64.nix;
    };
    musl = {
      aarch64-linux = import ./bootstrap-files/aarch64-musl.nix;
      armv6l-linux  = import ./bootstrap-files/armv6l-musl.nix;
      x86_64-linux  = import ./bootstrap-files/x86_64-musl.nix;
    };
  };

  # Try to find an architecture compatible with our current system. We
  # just try every bootstrap we’ve got and test to see if it is
  # compatible with or current architecture.
  getCompatibleTools = lib.foldl (v: system:
    if v != null then v
    else if localSystem.canExecute (lib.systems.elaborate { inherit system; }) then archLookupTable.${system}
    else null) null (lib.attrNames archLookupTable);

  archLookupTable = table.${localSystem.libc}
    or (abort "unsupported libc for the pure Linux stdenv");
  files = archLookupTable.${localSystem.system} or (if getCompatibleTools != null then getCompatibleTools
    else (abort "unsupported platform for the pure Linux stdenv"));
  in files
```

**Comment by trofi** [comment]
Also note that the actual error message (without the `--show-trace`) is one liner:

```
$ nix eval -f '<nixpkgs>' pkgsMusl.hello --argstr system i686-linux
error: evaluation aborted with the following error message: 'unsupported platform for the pure Linux stdenv'
(use '--show-trace' to show detailed location information)
```

**Comment by SuperSandro2000** [comment]
> I want to apply for the award for the worst error message with this one.

``Infinite recursion at undefined position`` is still better

> I used `--system i686-linux`, so that is not supported?

Seems like it for pure musl stdenv.

---

IMO this can be closed. There is nothing nix can do here except compacting the trace output. Everything else need to be done in nixpkgs.

**Comment by davidak** [comment]
> `Infinite recursion at undefined position` is still better

you are right. i forgot about that!

> > I used `--system i686-linux`, so that is not supported?
> 
> Seems like it for pure musl stdenv.

Yes, the error message should be more specific about the musl part. I also don't know if "pure Linux stdenv" makes sense. Is there an impure one?

> There is nothing nix can do here except compacting the trace output.

That part is already discussed in https://github.com/NixOS/nix/issues/7900.

> Everything else need to be done in nixpkgs.

I see. Please move the issue then.

**Comment by trofi** [comment]
> > > I used `--system i686-linux`, so that is not supported?
> > 
> > 
> > Seems like it for pure musl stdenv.
> 
> Yes, the error message should be more specific about the musl part. I also don't know if "pure Linux stdenv" makes sense. Is there an impure one?

Yes, "impure" stdenv is built from a compiler installed on user's system (as opposed to be coming from bootstrap binaries).

**Comment by trofi** [comment]
> nix eval -f '<nixpkgs>' pkgsMusl.hello --argstr system i686-linux --show-trace

And to state the obvious: `i686` as a primary build target is not even tested on hydra. It's absolutely expected that some things will not work. Normally that small subset of packages required for `i686` is built via `pkgsi686Linux.` build, like `pkgsi686Linux.tbb`.

**Comment by fricklerhandwerk** [comment]
This seems to be mostly a Nixpkgs issue, transferring.
